src/OpenRasta/UriTemplate.cs

```diff
@@ -375,25 +375,19 @@ public UriTemplateMatch Match(Uri baseAddress, string baseLeft,string[] baseSegm
       for (var i = 0; i < _segments.Count; i++)
       {
         var segment = candidateSegments[i];
+        var unescapedText = Uri.UnescapeDataString(segment);
 
-        var candidateSegment = new
-        {
-          Text = segment,
-          UnescapedText = Uri.UnescapeDataString(segment),
-          ProposedSegment = _segments[i]
-        };
-
-        candidateSegments[i] = candidateSegment.Text;
+        candidateSegments[i] = segment;
 
-        switch (candidateSegment.ProposedSegment.Type)
+        switch (_segments[i].Type)
         {
           case SegmentType.Literal when
-            string.CompareOrdinal(candidateSegment.ProposedSegment.Text, candidateSegment.UnescapedText) != 0:
+            string.CompareOrdinal(_segments[i].Text, unescapedText) != 0:

```

Comment:
Not your code but this could be simplified to `!string.Equals(_segments[i].Text, unescapedText, StringComparison.OrdinalIgnoreCase)`.